File: sem2_lab4.2/sem2_lab42.cpp
#include <iostream>
#include <random>
#include <time.h>
#include <vector>

using namespace std;

struct number
{
    int value;
    int place;
};

int first_digit(int number)
{
    while (number > 9)
    {
        number /= 10;
    }
    return number;
}

void nums_order(vector<number> &numbers, int past_num, int num_in_seq, int &size)
{
    bool numAdded = false;

    for (int i = 0; i < size; i++)
    {
        if (past_num == -1 || (numbers[i].place == -1 && first_digit(numbers[i].value) == past_num % 10))
        {
            numbers[i].place = num_in_seq;
            nums_order(numbers, numbers[i].value, num_in_seq + 1, size);
            numbers[i].place = -1;
            numAdded = true;
        }
    }

    if (!numAdded && num_in_seq > 2)
    {
        for (int i = 0; i < num_in_seq; i++)
        {
            for (int j = 0; j < size; j++)
            {
                if (numbers[j].place == i)
                {
                    cout << numbers[j].value << " ";
                    break;
                }
            }
        }
        cout << endl;
    }
}

int main()
{

    int size = 9;
    vector<number> numbers(size);

    srand(time(0));

    cout << "Generated nums: ";
    for (int i = 0; i < size; i++)
    {
        numbers[i].value = rand() % 100;
        numbers[i].place = -1;
        cout << numbers[i].value << " ";
    }
    cout << "\n\n";
    cout << "Combinations: " << endl;
    nums_order(numbers, -1, 0, size);
}

File: sem2_lab4.2/sem2_lab42.py
import random

class number:
    value = 0
    place = 0 
    def __init__(self, value, place):
        self.value = value
        self.place = place
    
def first_digit(number):
    while number > 9:
        number //= 10
    return number

def nums_order(numbers, past_num, num_in_seq, size):
    numAdded = False

    i = 0
    for i in range(0,size):
        if (past_num == -1 or (numbers[i].place == -1 and first_digit(numbers[i].value) == past_num % 10)):
            numbers[i].place = num_in_seq
            nums_order(numbers, numbers[i].value, num_in_seq + 1, size)
            numbers[i].place = -1
            numAdded = True
        
        if (not numAdded and num_in_seq > 2):
            i = 0
            j = 0
            for i in range(0,num_in_seq):
                for j in range(0,size):
                    if numbers[j].place == i:
                        print(numbers[j].value, end = ' ')
            print('')

size = 10
numbers = []

print("Generated nums: ")
i = 0
for i in range(0,size):
    numbers.append(number(random.randint(0, 101),-1))
    print(numbers[i].value, end = " ")

print("")
print("Combinations: ")
nums_order(numbers, -1, 0, size)